Remove 8.8.8.8 and instead determine the local IP for each packet

Flole998 opened the pull request:
Still needs testing

Fixes: #1634

Flole998 committed:
Remove 8.8.8.8 and instead determine the local IP for each packet

Flole998 commented on src/satip/server.c:
```diff
@@ -988,18 +1000,10 @@ static void satip_server_init_common(const char *prefix, int announce)
-    http_server_ip = strdup(satip_server_bindaddr ?: http_ip);
-    http_server_port = ntohs(IP_PORT(http));
+  if (http_server_port == 0) {
+    tcp_server_onall(http_server);
```
Intended.

Flole998 commented on src/satip/server.c:
```diff
@@ -420,8 +430,10 @@ CONFIGID.UPNP.ORG: 0\r\n"
 
+  char* own_server_ip = http_server_ip;
+
   snprintf(buf, sizeof(buf), MSG, UPNP_MAX_AGE,
```
Intended

Flole998 commented on src/satip/rtsp.c:
```diff
@@ -1884,7 +1884,8 @@ void satip_server_rtsp_init
   }
   s = rtsp_ip;
-  rtsp_ip = strdup(bindaddr);
+  if(bindaddr != NULL)
```
Thats.... uhmmm... literally what I am doing?

dave-p commented:
I tried to test this PR as a possible fix for "Sat>IP Server only working when internet is connected" in the Forum, however the code does not compile when committed against current Master. The errors are:

```
src/tcp.c:1021:1: error: no previous prototype for ‘tcp_server_bound’ [-Werror=missing-prototypes]
 1021 | tcp_server_bound ( void *server, struct sockaddr_storage *bound, int family )
      | ^~~~~~~~~~~~~~~~
src/tcp.c: In function ‘tcp_server_bound’:
src/tcp.c:1044:7: error: implicit declaration of function ‘tcp_default_ip_addr’ [-Wimplicit-function-declaration]
 1044 |   if (tcp_default_ip_addr(bound, family) < 0)
      |       ^~~~~~~~~~~~~~~~~~~

      
src/udp.c:494:1: error: no previous prototype for ‘udp_write_fill_source’ [-Werror=missing-prototypes]
  494 | udp_write_fill_source( udp_connection_t *uc, const void *buf, size_t len,
      | ^~~~~~~~~~~~~~~~~~~~~
src/udp.c: In function ‘udp_write_queue’:
src/udp.c:572:10: error: too few arguments to function ‘udp_write_queue_fill_source’
  572 |   return udp_write_queue_fill_source(uc, q, storage);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from src/udp.c:23:
src/udp.h:70:1: note: declared here
   70 | udp_write_queue_fill_source( udp_connection_t *uc, htsbuf_queue_t *q,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~


src/satip/server.c: In function ‘satips_upnp_send_byebye’:
src/satip/server.c:338:5: error: too few arguments to function ‘upnp_send’
  338 |     upnp_send(&q, NULL, attempt * 11, 1);
      |     ^~~~~~~~~
In file included from src/satip/server.c:21:
src/upnp.h:42:6: note: declared here
   42 | void upnp_send(htsbuf_queue_t *q, struct sockaddr_storage *storage,
      |      ^~~~~~~~~
src/satip/server.c: In function ‘satip_server_init_common’:
src/satip/server.c:1005:78: error: ‘null’ undeclared (first use in this function)
 1005 |     http_server_ip = satip_server_bindaddr ? strdup(satip_server_bindaddr) : null;
      |                                                                              ^~~~
In file included from src/tvheadend.h:28,
                 from src/satip/server.c:20:
src/satip/server.c:1006:39: error: ‘tcp_server_t’ undeclared (first use in this function); did you mean ‘tcp_server_init’?
 1006 |     http_server_port = ntohs(IP_PORT((tcp_server_t *)http_server));
      |                                       ^~~~~~~~~~~~
src/satip/server.c:1006:30: note: in expansion of macro ‘IP_PORT’
 1006 |     http_server_port = ntohs(IP_PORT((tcp_server_t *)http_server));
      |                              ^~~~~~~
src/satip/server.c:1006:53: error: expected expression before ‘)’ token
 1006 |     http_server_port = ntohs(IP_PORT((tcp_server_t *)http_server));
      |                                                     ^
src/satip/server.c:995:8: error: unused variable ‘http_ip’ [-Werror=unused-variable]
  995 |   char http_ip[128];
      |        ^~~~~~~
src/satip/server.c:994:27: error: unused variable ‘http’ [-Werror=unused-variable]
  994 |   struct sockaddr_storage http;
      |                           ^~~~
```
tcp.c is simple to fix (delete function tcp_server_bound in tcp.c and tcp.h) however I don't understand the code enough to fix the rest.

chkuendig commented:
did manage to fix and compile this (on top of #1679) but I'm not sure it works. My issue is that the code at https://github.com/tvheadend/tvheadend/blob/1134ca16a7acfc76c5a5c2c046a80fb771e5ad48/src/udp.c#L510-L516

Outputs 
`upnp: Got source IP address: 239.255.255.250`

Looks like it's bound to the multicast IP so it doesn't have the local IP. @Flole998 Did this work for you?

Flole998 commented:
@chkuendig I haven't had time to play with this again, however the source should never be a multicast address, so this is probably the destination that ends up in there? That is not correct of course and will then prevent it from working properly.

As I am waiting for something else to compile right now I can have a look at this again right now.

chkuendig commented:
I think the connection we are accessing is created at https://github.com/tvheadend/tvheadend/blob/1134ca16a7acfc76c5a5c2c046a80fb771e5ad48/src/upnp.c#L133-L135 and passed to the udp code at https://github.com/tvheadend/tvheadend/blob/1134ca16a7acfc76c5a5c2c046a80fb771e5ad48/src/upnp.c#L190 which would explain why this IP ends up here.

Flole998 commented:
It should not be getting the local IP but the remote IP though when a packet is received.... Let me first fix the issues that prevent this from compiling.....

Flole998 commented:
Okay this should be working, but I likely introduced some security issues. I will have to look at it again and fix those.

chkuendig commented:
This seems to work. I'll test it for a while (for HDHomerun Server Emulation - I don't have a SAT>IP setup) and see how it goes.

chkuendig commented on src/udp.c:
```diff
@@ -490,6 +490,98 @@ udp_close( udp_connection_t *uc )
+    len += strlen(inet_ntoa(local_addr.sin_addr)) - 2;
+    char* data = malloc(len + 1);
+
+    len = snprintf(data, len + 1, buf, inet_ntoa(local_addr.sin_addr));
```
Theres a bug here. 

snprintf returns "The number of characters that would have been written if n had been sufficiently large". This number can be larger than the size of 
`data` (and in my tests often is - I think because `buf` isnt a null terminated string). I think whats needed is to check that the return is >0 but ignore the specific value.

Alternatively, if `buf` were null-terminated, the return value should always be the same as `len`.

Flole998 replied:
This needs to be revisited anyways, as inet_ntoa will fail on IPv6, so this is not quite right yet and a few adjustments are still needed. I will keep this in mind, thanks!

chkuendig reviewed (commented):
Added a comment, in my tests some upnp messages contain random data from memory (or the process crashes if memory isnt accessible)

sgllama commented:
How does this fix work when you have multiple IP addresses available? Either because of multiple NICs (WiFi + wired, two wired...) or because you have a NIC that is multihomed?

Hopefully I missed something, but the isn't the only solution that gets over all the issues to simply have an edit field in the Web UI that allows you to type in the IP? Or a combobox, pre-filled with all the available options if you want to be really nice to the user?

(I note that the original code, talking to 8.8.8.8, was trying to not only find a local IP ''but'' also the local IP that has direct access to the global Internet, which is - wonderfully questionable behaviour if you do have multiple IPs, ''but'' it does mean that the author didn't have to worry about filtering out 127.0.x.x out when listing the interfaces!)

Flole998 commented:
> How does this fix work when you have multiple IP addresses available? Either because of multiple NICs (WiFi + wired, two wired...) or because you have a NIC that is multihomed?

It will use your systems routing table.



> Hopefully I missed something, but the isn't the only solution that gets over all the issues to simply have an edit field in the Web UI that allows you to type in the IP?

The possibility to manually specify it already exists, but this is the "fallback" for cases where this hasn't been done. If you are behind a firewall with NAT you might even want to specify something completely different here.



> I note that the original code, talking to 8.8.8.8, was trying to not only find a local IP ''but'' also the local IP that has direct access to the global Internet, which is - wonderfully questionable behaviour if you do have multiple IPs,

Correct, and it stalled if 8.8.8.8 was not reachable....



> ''but'' it does mean that the author didn't have to worry about filtering out 127.0.x.x out when listing the interfaces!

And with this change the system Routing table is used to find out how the packet should be sent, and then the source address of the packet is told to the client to connect to. This is as good as the automatic stuff can get IMO, and you still have the manual override if this is not working for whatever reason.

Flole998 committed:
Fix UDP source IP determination logic for UPnP/SSDP

This refactors the attempt to determine the local source IP without connecting to 8.8.8.8. Instead of cloning a socket in udp.c and injecting the IP unsafely into formatted strings, it now determines the appropriate local IP via a dummy UDP connect to the destination IP directly in satip/server.c and cleanly substitutes it into the message buffer before queueing it to upnp_send.

Flole998 committed:
Log error and abort instead of using 0.0.0.0 when source IP resolution fails

Flole998 committed:
Include arpa/inet.h for inet_pton in satip/server.c
